File: tutorial/lect2-sigProc/SigProcERSP.m
%% initialise the matlab paths
try; cd(fileparts(mfilename('fullpath')));catch; end;
try;
   run ../../matlab/utilities/initPaths.m
catch
   msgbox({'Please change to the directory where this file is saved before running the rest of this code'},'Change directory'); 
end

%% load the data this contains
% X - [ channels x time x epochs ] raw EEG data
% Y - [ epochs x 1 ] class labels for X's epochs
% fs- [int] sample rate of X
% Cnames - {str nCh x 1} cell array of strings for the names of X's channels
% Cpos - [2 x nCh] x-y positions of each of X's channels (for plotting)
load('ERSPdata'); 
times=(1:size(X,2))/fs; yvals=times; ylab='time (s)';

%% plot the first 3 epochs of the data in a topo-graphic multi-plot
figure(1);set(gcf,'Name','Single Trials');clf;image3d(X(:,:,1:3),1,'plotPos',Cpos,'Xvals',Cnames,'Yvals',yvals,'ylabel',ylab,'disptype','plot','ticklabs','sw');
%% plot the class averages
erp = cat(3,mean(X(:,:,Y>0),3),mean(X(:,:,Y<=0),3));
figure(2);set(gcf,'Name','Class Average');clf;image3d(erp,1,'plotPos',Cpos,'Xvals',Cnames,'Yvals',yvals,'ylabel',ylab,'Zvals',{'pos','neg'},'disptype','plot','ticklabs','sw');
zoomplots; % allow interactive zooming of the plots to see better what's going on

%% 1) detrend and try again
fprintf('1) Detrend\n');
X=detrend(X,2);
updateSigProcPlots; %Update the plots

%% 2) remove bad channels
fprintf('2) bad channel removal, ');
% compute and plot the power in each channel..
chpow = sqrt(sum(X(:,:).^2,2)./size(X,2)./size(X,3)); % mean amplitude per channel
figure(3); clf;plot(chpow,'*');ylabel('amplitude (uV)');title('channel power');
set(gca,'xtick',1:2:numel(Cnames),'xticklabel',Cnames(1:2:end)); % put on the channel names..

%% automatically identify the bad channels
[badch,feat,threshs] = idOutliers(X,1,2.5);
% show the bad channels on the power plot
figure(3);hold on;plot(find(badch),ones(size(find(badch)))*threshs(end),'r*','markersize',10); 
legend('ch-power','bad-ch');

% remove them from the data
X=X(~badch,:,:);
% don't forget to update the channel info too!
Cpos=Cpos(:,~badch);
Cnames=Cnames(~badch);
updateSigProcPlots; %Update the plots

%% 3) re-reference - CAR
fprintf('3) CAR\n');
% plot the common average signal, for all trials
figure(3);clf; plot(shiftdim(mean(X,1)));title('Per trial CAR signal');
X =X - repmat(mean(X,1),[size(X,1),1,1]);
% alt faster version: X = repop(X,'-',mean(X,1));
updateSigProcPlots; %Update the plots

%% 3.5) remove bad epochs
fprintf('3.5) bad trial removal');
% plot the epoch/trial powers
eppow = sqrt(sum(reshape(X,[],size(X,3)).^2,1)./size(X,1)./size(X,2));
figure(3); clf; plot(eppow,'*');title('Trial power');
% automatically identify the bad epochs
[badep,feat,threshs] = idOutliers(X,3,3);
% show them on the power plot
figure(3);hold on;plot(find(badep),ones(size(find(badep)))*threshs(end),'r*','markersize',10); 
legend('trial-power','bad-trials');

%% store the bad examples and plot them
Xbad = X(:,:,badep); % store bad example to plot it..
% plot the bad epochs
figure(4); clf;image3d(Xbad,1,'plotPos',Cpos,'Xvals',Cnames,'Zvals',find(badep),'disptype','plot','ticklabs','sw','ylabel','samples','zlabel','epoch','clabel','uV');suptitle('Bad epochs');

% remove the bad epochs from the data
X=X(:,:,~badep);
Y=Y(~badep); % don't forget to update labels!
updateSigProcPlots; %Update the plots

%% 4) welch to convert to power spectral density
fprintf('4) Welch\n');
[X,wopts,winFn]=welchpsd(X,2,'width_ms',500,'fs',fs);
freqs=0:(1000/500):fs; % position of the frequency bins
yvals=freqs; ylab='freq (Hz)';
updateSigProcPlots; %Update the plots

%% 5) sub-select the range of frequencies we care about
freqband = [8 28];
fidx=[];
fprintf('5) Select frequencies\n');
[ans,fidx(1)]=min(abs(freqs-freqband(1))); % lower frequency bin
[ans,fidx(2)]=min(abs(freqs-freqband(2))); % upper frequency bin
X=X(:,fidx(1):fidx(2),:); % sub-set to the interesting frequency range
yvals=yvals(fidx(1):fidx(2)); % update so plots use correct info
updateSigProcPlots; %Update the plots

%% 6) train classifier
fprintf('6) train classifier\n');
[clsfr, res]=cvtrainLinearClassifier(X,Y,[],10);
% show the classifier structure
clsfr
%% Plot the trained classifier weight vector (only for linear classifiers)
figure(3);clf;image3d(clsfr.W,1,'plotPos',Cpos,'Xvals',Cnames,'Yvals',yvals,'disptype','plot','ticklabs','sw','ylabel','freq (hz)','zlabel','clsfr','clabel','weight'); suptitle('Classifier weights');

%---------------------------------------------------------------------------------------------------
%%7) use train_ersp_clsfr which does all this for you!
[clsfr,res]=train_ersp_clsfr(X,Y,'freqband',[8 28],'ch_pos',Cnames,'fs',fs);

File: tutorial/lect2-sigProc/SigProcERSP.py
#!/usr/bin/env python
import os, sys
# Path of the folder containing the buffer client                                              
try:
    pydir=os.path.dirname(__file__)
except:
    pydir=os.getcwd()

sigProcPath = '../../python/signalProc'
sys.path.append(os.path.join(os.path.abspath(pydir),sigProcPath))
import preproc
import linear
import numpy as np
plottingPath= '../../python/plotting'
sys.path.append(os.path.join(os.path.abspath(pydir),plottingPath))
from image3d import *
from scipy.io import loadmat

#load the datafile, and extract the variables                                                  
data=loadmat('ERSPdata.mat')
X     =data['X']; print("X= [channels x timepoints x trials]");print(X.shape)
Y     =data['Y'].reshape(-1); print("Y=[trials]");print(Y.shape)
fs    =float(data['fs'][0])
Cnames=data['Cnames'].reshape(-1); 
Cnames=np.array([item for sublist in Cnames for item in sublist]) #flatten the list of lists of names and make array
print("Channel names : ");print(Cnames)
Cpos  =data['Cpos']; print("Cpos= [ 3 x channels]");print(Cpos.shape);


# plot the data                                                                                
image3d(X[:,:,1:3],0,plotpos=Cpos);

# plot the class averages
erp=np.stack((X[:,:,Y>0].mean(2),X[:,:,Y<=0].mean(2)),2) #compute the ERP
image3d(erp,0,plotpos=Cpos,xvals=Cnames); # plot the ERPs

# define some utility functions to simplify plotting of data
ylabel='time (s)'
yvals =np.arange(0,X.shape[1])/fs  # element labels for 2nd dim of X

def plotTrials(X,trls):
    'Plot the single-trial data'
    global Cnames, Cpos, ylabel, yvals
    image3d(X[:,:,1:3],0,plotpos=Cpos,xvals=Cnames,ylabel=ylabel,yvals=yvals)

def plotERP(X,Y):
    'plot the class averages'
    global Cnames, Cpos, ylabel, yvals
    erp=np.stack((X[:,:,Y>0].mean(2),X[:,:,Y<=0].mean(2)),2) #compute the ERP
    image3d(erp,0,plotpos=Cpos,xvals=Cnames,ylabel=ylabel,yvals=yvals) # plot the ERPs

def updatePlots():
    global X,Y
    # Plot the raw data
    fig=plt.figure(1,figsize=(12, 12))
    plotTrials(X,range(3)) # single trial

    fig=plt.figure(2,figsize=(12, 12))
    plotERP(X,Y) # class averages

# plot a subset of single-trials
plt.figure(1,figsize=(12, 12));
plotTrials(X,range(3));


# plot the grand average's per condition
plt.figure(2,figsize=(12, 12))
plotERP(X,Y);


#-------------------------------------------------------------------
#  Run the standard pre-processing and analysis pipeline
# 1: detrend
X        = preproc.detrend(X)
updatePlots();


# 2: bad-channel removal
goodch, badch = preproc.outlierdetection(X);
X = X[goodch,:,:];
Cnames=Cnames[goodch];
Cpos=Cpos[:,goodch];
updatePlots()


# 3: apply spatial filter
X        = preproc.spatialfilter(X,type='car')
updatePlots();


# Map to frequency domain, only keep the positive frequencies
X,freqs = preproc.powerspectrum(X,dim=1,fSample=fs)
yvals = freqs; # ensure the plots use the right x-ticks
ylabel='freq (Hz)'
updatePlots()


# 5 : select the frequency bins we want
X,freqIdx=preproc.selectbands(X,dim=1,band=[8,10,28,30],bins=freqs)
freqs=freqs[freqIdx]
yvals=freqs
updatePlots()


# 6 : bad-trial removal, trials in dim=2
goodtr, badtr = preproc.outlierdetection(X,dim=2)
X = X[:,:,goodtr]
Y = Y[goodtr]
updatePlots()


# 7: train linear least squares classifier, with cross-validation
import sklearn
clsfr = sklearn.linear_model.RidgeCV(store_cv_values=True)
X2d = np.reshape(X,(-1,X.shape[2])).T # sklearn needs x to be [nTrials x nFeatures]
clsfr.fit(X2d,Y)
print("MSSE=%g"%np.mean(clsfr.cv_values_))

# plot the classifier weight vector
W = clsfr.coef_
W = np.reshape(clsfr.coef_,(X.shape[0],X.shape[1])) 
image3d(W,0,plotpos=Cpos,xvals=Cnames,yvals=freqs);
